File: examples/python/patterns/saga-pattern/saga-pattern-parallel.py
from aws_durable_execution_sdk_python import (
    BatchResult,
    DurableContext,
    StepContext,
    durable_execution,
    durable_step,
)

@durable_step
def reserve_inventory(ctx: StepContext, order_id: str) -> dict:
    return {"id": f"RES-{order_id}"}


@durable_step
def cancel_reservation(ctx: StepContext, reservation_id: str) -> None:
    print(f"Reservation {reservation_id} cancelled")


@durable_step
def verify_address(ctx: StepContext, address: dict) -> dict:
    # Stateless — no compensation needed
    return {"valid": True}


@durable_step
def charge_payment(ctx: StepContext, amount: float) -> dict:
    return {"id": f"PAY-{amount}"}


@durable_step
def refund_payment(ctx: StepContext, payment_id: str) -> None:
    print(f"Payment {payment_id} refunded")


@durable_step
def create_shipment(ctx: StepContext, order_id: str) -> dict:
    return {"tracking_id": f"TRACK-{order_id}"}

@durable_execution
def handler(event: dict, context: DurableContext) -> dict:
    compensations = []

    try:
        order_id = event["order_id"]
        amount   = event["amount"]
        address  = event["address"]

        # Reserve inventory AND verify address in parallel
        pre_checks: BatchResult = context.parallel(
            [
                lambda ctx: ctx.step(reserve_inventory(order_id)),
                lambda ctx: ctx.step(verify_address(address)),
            ],
            name="pre-checks",
        )

        reservation, address_verification = pre_checks.get_results()

        compensations.append(("cancel-reservation", cancel_reservation, reservation["id"]))

        # Stop execution if address is invalid
        if not address_verification.get("valid"):
            raise ValueError("Invalid shipping address")

        payment = context.step(charge_payment(amount))
        compensations.append(("refund-payment", refund_payment, payment["id"]))

        shipment = context.step(create_shipment(order_id))

        return {"success": True, "tracking_id": shipment["tracking_id"]}

    except Exception as error:
        context.logger.error("Order failed, running compensations", error)

        for name, comp_step, resource_id in reversed(compensations):
            try:
                context.step(comp_step(resource_id), name=name)
            except Exception as comp_error:
                context.logger.error(f"Compensation failed: {name}", comp_error)

        raise error

File: examples/typescript/patterns/saga-pattern/saga-pattern-parallel.ts
import { withDurableExecution, DurableContext } from '@aws/durable-execution-sdk-js';

export const handler = withDurableExecution(async (event: any, context: DurableContext) => {
  const compensations: Array<{ name: string; fn: () => Promise<void> }> = [];

  try {
    // Reserve inventory AND verify address in parallel
    const preChecks = await context.parallel('pre-checks', [
      async (ctx) => ctx.step('reserve-inventory', async () =>
        reserveInventory(event.order_id, event.items)
      ),
      async (ctx) => ctx.step('verify-address', async () =>
        verifyAddress(event.address)
      ),
    ]);

    const [reservation, isValidAddress] = preChecks.getResults() as [{ id: string }, { valid: boolean }];

    // Only reserve-inventory needs a compensation because verify-address is stateless
    compensations.push({
      name: 'cancel-reservation',
      fn: () => cancelReservation(reservation.id)
    });

    // Stop execution if address is invalid, catch block will cancel reservation
    if (!isValidAddress.valid) {
      throw new Error('Invalid shipping address');
    }

    const payment = await context.step('charge-payment', async () =>
      chargePayment(event.payment_method, event.amount)
    );
    compensations.push({
      name: 'refund-payment',
      fn: () => refundPayment(payment.id)
    });

    const shipment = await context.step('create-shipment', async () =>
      createShipment(event.order_id, event.address)
    );

    return { success: true, tracking_id: shipment.tracking_id };

  } catch (error) {
    context.logger.error('Order failed, running compensations', { error });

    for (const comp of compensations.reverse()) {
      try {
        await context.step(comp.name, async () => comp.fn());
      } catch (compError) {
        context.logger.error(`Compensation failed: ${comp.name}`, compError);
      }
    }

    throw error;
  }
});

// Mock APIs for demonstration purposes

async function reserveInventory(orderId: string, items: any[]): Promise<{ id: string }> {
  return { id: `RES-${orderId}` };
}

async function cancelReservation(reservationId: string): Promise<void> {
  console.log(`Reservation ${reservationId} cancelled`);
}

async function verifyAddress(address: any): Promise<{ valid: boolean }> {
  return { valid: true };
}

async function chargePayment(method: any, amount: number): Promise<{ id: string }> {
  return { id: `PAY-${amount}` };
}

async function refundPayment(paymentId: string): Promise<void> {
  console.log(`Payment ${paymentId} refunded`);
}

async function createShipment(orderId: string, address: any): Promise<{ tracking_id: string }> {
  return { tracking_id: `TRACK-${orderId}` };
}
